compiler/modules/front/expr.bal

```diff
@@ -1398,15 +1416,18 @@ function codeGenCheckingTerminator(bir:BasicBlock bb, s:CheckingKeyword checking
 function codeGenFunctionCallExpr(ExprContext cx, bir:BasicBlock bb, s:FunctionCallExpr expr) returns CodeGenError|ExprEffect {
     string? prefix = expr.prefix;
     bir:FunctionRef func;
-    bir:Register? funcRegister = ();
+    bir:AssignTmpRegister? funcRegister = ();
     if prefix == () {
         var ref = cx.lookupLocalVarRef(expr.funcName, expr.qNamePos);
         if ref is bir:FunctionRef {
             func = ref;
         }
         else if ref is Binding {
-            func = check functionRefFromRegister(cx, ref.reg, expr.qNamePos);
-            funcRegister = ref.reg;
+            var [semType, atom] = check selectFunctionInherentType(cx, ref.reg.semType, expr);
```

Comment:
I'm not sure this is the right semantics for function call. Have we looked at how the Covariance/Contravariance paper types functional call?  The spec doesn't have any concept of function inherent type.

Reply:
The problem I am trying to solve here is 
```
import ballerina/io;
type F function(int) returns int;
type F2 function(int, int) returns int;

public function main() {
    F|F2 f = f1;
    if f is F {
        io:println(f(1));
    }
    else {
        io:println(f(1, 2)); // <- what is the type of f
    }
}

function f1(int i) returns int {
  return i;
}
```
Currently at the point I have marked we get the type as a `diff` of two types (union and `F`), but we need a single function type (at least I think we need one to generate code).

Reply:
Fixed